# Data Cleaning Notebook

In [1]:
# Import the required libraries
import pandas as pd
import numpy as np

# set up pandas to display floats in a more human friendly way
pd.options.display.float_format = '{:,.2f}'.format

### Original training data files downloaded from data source were split between values and labels. 
#### Step 1: Conduct data cleaning USING FUNCTIONS on the Training and Test Files
#### Step 2: Combine the Training values and labels into a single file for use in model training.

In [2]:
# TRAINING DATA
train_values_raw = pd.read_csv('../data/original/TrainingSetValues/4910797b-ee55-40a7-8668-10efd5c1b960.csv')
train_values_raw.head(3)

id  amount_tsh date_recorded        funder  gps_height     installer  \
0  69572    6,000.00    2011-03-14         Roman        1390         Roman   
1   8776        0.00    2013-03-06       Grumeti        1399       GRUMETI   
2  34310       25.00    2013-02-25  Lottery Club         686  World vision   

   longitude  latitude     wpt_name  num_private  ... payment_type  \
0      34.94     -9.86         none            0  ...     annually   
1      34.70     -2.15     Zahanati            0  ...    never pay   
2      37.46     -3.82  Kwa Mahundi            0  ...   per bucket   

  water_quality quality_group      quantity  quantity_group  \
0          soft          good        enough          enough   
1          soft          good  insufficient    insufficient   
2          soft          good        enough          enough   

                 source           source_type  source_class  \
0                spring                spring   groundwater   
1  rainwater harvesting  rainwater harvesting       surface   
2                   dam                   dam       surface   

               waterpoint_type waterpoint_type_group  
0           communal standpipe    communal standpipe  
1           communal standpipe    communal standpipe  
2  communal standpipe multiple    communal standpipe  

[3 rows x 40 columns]

In [3]:
train_values_raw.describe()

id  amount_tsh  gps_height  longitude  latitude  num_private  \
count 59,400.00   59,400.00   59,400.00  59,400.00 59,400.00    59,400.00   
mean  37,115.13      317.65      668.30      34.08     -5.71         0.47   
std   21,453.13    2,997.57      693.12       6.57      2.95        12.24   
min        0.00        0.00      -90.00       0.00    -11.65         0.00   
25%   18,519.75        0.00        0.00      33.09     -8.54         0.00   
50%   37,061.50        0.00      369.00      34.91     -5.02         0.00   
75%   55,656.50       20.00    1,319.25      37.18     -3.33         0.00   
max   74,247.00  350,000.00    2,770.00      40.35     -0.00     1,776.00   

       region_code  district_code  population  construction_year  
count    59,400.00      59,400.00   59,400.00          59,400.00  
mean         15.30           5.63      179.91           1,300.65  
std          17.59           9.63      471.48             951.62  
min           1.00           0.00        0.00               0.00  
25%           5.00           2.00        0.00               0.00  
50%          12.00           3.00       25.00           1,986.00  
75%          17.00           5.00      215.00           2,004.00  
max          99.00          80.00   30,500.00           2,013.00

In [4]:
# TEST DATA 
test_values_raw = pd.read_csv('../data/original/TestSetValues/702ddfc5-68cd-4d1d-a0de-f5f566f76d91.csv')
test_values_raw.head(3)

id  amount_tsh date_recorded                  funder  gps_height  \
0  50785        0.00    2013-02-04                    Dmdd        1996   
1  51630        0.00    2013-02-04  Government Of Tanzania        1569   
2  17168        0.00    2013-02-01                     NaN        1567   

  installer  longitude  latitude                 wpt_name  num_private  ...  \
0      DMDD      35.29     -4.06  Dinamu Secondary School            0  ...   
1       DWE      36.66     -3.31                  Kimnyak            0  ...   
2       NaN      34.77     -5.00           Puma Secondary            0  ...   

  payment_type water_quality quality_group      quantity  quantity_group  \
0    never pay          soft          good      seasonal        seasonal   
1    never pay          soft          good  insufficient    insufficient   
2    never pay          soft          good  insufficient    insufficient   

                 source           source_type  source_class  \
0  rainwater harvesting  rainwater harvesting       surface   
1                spring                spring   groundwater   
2  rainwater harvesting  rainwater harvesting       surface   

      waterpoint_type waterpoint_type_group  
0               other                 other  
1  communal standpipe    communal standpipe  
2               other                 other  

[3 rows x 40 columns]

In [5]:
test_values_raw.describe()

id  amount_tsh  gps_height  longitude  latitude  num_private  \
count 14,850.00   14,850.00   14,850.00  14,850.00 14,850.00    14,850.00   
mean  37,161.97      322.83      655.15      34.06     -5.68         0.42   
std   21,359.36    2,510.97      691.26       6.59      2.94         8.17   
min       10.00        0.00      -57.00       0.00    -11.56         0.00   
25%   18,727.00        0.00        0.00      33.07     -8.44         0.00   
50%   37,361.50        0.00      344.00      34.90     -5.05         0.00   
75%   55,799.75       25.00    1,308.00      37.20     -3.32         0.00   
max   74,249.00  200,000.00    2,777.00      40.33     -0.00       669.00   

       region_code  district_code  population  construction_year  
count    14,850.00      14,850.00   14,850.00          14,850.00  
mean         15.14           5.63      184.11           1,289.71  
std          17.19           9.67      469.50             955.24  
min           1.00           0.00        0.00               0.00  
25%           5.00           2.00        0.00               0.00  
50%          12.00           3.00       20.00           1,986.00  
75%          17.00           5.00      220.00           2,004.00  
max          99.00          80.00   11,469.00           2,013.00

## Column and Row Information
* 40 columns/features in the raw data TRAINING and TEST
* 59,400 rows in the TRAINING data
* 14,850 rows in the TEST data

In [6]:
train_labels_raw = pd.read_csv('../data/original/TrainingSetLabels/0bf8bc6e-30d0-4c50-956a-603fc693d966.csv')
print(train_labels_raw.info())
train_labels_raw.head(2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   id            59400 non-null  int64 
 1   status_group  59400 non-null  object
dtypes: int64(1), object(1)
memory usage: 928.2+ KB
None


id status_group
0  69572   functional
1   8776   functional

## Original Data Column Descriptions

* id - Numeric identifer for the waterpoint
* amount_tsh - Total static head (amount water available to waterpoint)
* date_recorded - The date the row was entered
* funder - Who funded the well
* gps_height - Altitude of the well
* installer - Organization that installed the well
* longitude - GPS coordinate
* latitude - GPS coordinate
* wpt_name - Name of the waterpoint if there is one
* num_private -
* basin - Geographic water basin
* subvillage - Geographic location
* region - Geographic location, NOTE: Hierarchy is Region > LGA > Ward
* region_code - Geographic location (coded)
* district_code - Geographic location (coded)
* lga - Geographic location
* ward - Geographic location
* population - Population around the well
* public_meeting - True/False
* recorded_by - Group entering this row of data
* scheme_management - Who operates the waterpoint
* scheme_name - Who operates the waterpoint
* permit - If the waterpoint is permitted
* construction_year - Year the waterpoint was constructed
* extraction_type - The kind of extraction the waterpoint uses
* extraction_type_group - The kind of extraction the waterpoint uses
* extraction_type_class - The kind of extraction the waterpoint uses
* management - How the waterpoint is managed
* management_group - How the waterpoint is managed
* payment - What the water costs
* payment_type - What the water costs
* water_quality - The quality of the water
* quality_group - The quality of the water
* quantity - The quantity of water
* quantity_group - The quantity of water
* source - The source of the water
* source_type - The source of the water
* source_class - The source of the water
* waterpoint_type - The kind of waterpoint
* waterpoint_type_group - The kind of waterpoint



## Data cleaning steps:
* __Duplicate check__ 

* __Address Null Values__

* __Address Zeros in Numeric Values__

* __String type normalization__
 
* __New column/feature creation__


* __Note:__ No Data type conversions needed

### Duplicate Check

In [7]:
# Functions for Duplicate checks 
def has_exact_dups(df):
    dups = df[df.duplicated()]
    return len(dups) > 0

def has_identifier_dups(df, col_name='id'):
    num_rows = df.shape[0]
    num_ids = len(df[col_name].unique())
    return num_ids != num_rows

In [8]:
#Dup checking
print(has_exact_dups(train_values_raw))
print(has_identifier_dups(train_values_raw))
print(has_exact_dups(test_values_raw))
print(has_identifier_dups(test_values_raw))

False
False
False
False


### There are no exact duplicate rows nor duplicate identifers in the Training or Test datasets

In [9]:
# Make a deep copy and do procesing on this dataframe moving forward
# Deep copy has own copy of data and index.
train_values_processed = train_values_raw.copy(deep=True)
test_values_processed = test_values_raw.copy(deep=True)

In [10]:
#Null handler functions
def get_null_stats(df, col_name):
    print(df[col_name].unique())
    print(len(df[col_name].unique()))
    print(df[col_name].isna().sum())


def handle_nulls_inplace(df, cols_to_fill):
    for item in cols_to_fill:
        for key, value in item.items():
            df[key].fillna(value, inplace=True)

### Handling Null Values
#### TRAINING and TEST columns with Null values:
* funder : set to 'unknown' when null
* installer : set to 'unknown' when null
* subvillage : set to 'unknown' when null
* public_meeting : set to True when null
    * Training dataset: only 5055 out of 59,400 were False. 51,011 out of 59,4000 were True. 
    * Test dataset: only 1291 out of 14,850 were False, 12,738 out of 14,850 were True.
    * As the vast majority of both Training and Test waterpoints have public_meeting of True, use True to replace all nulls.
* scheme_management : set to 'unknown' when null
* scheme_name : set to 'unknown' when null
* permit : set to True when nulls
    * Training dataset: 38,852 out of 59,4000 were True. 17,492 out of 59,4000 where False. 
    * Test dataset:  9754 out of 14,850 were True. 4359 out of 14,850 were False
    * As the majority of both Training and Test waterpoints have public_meeting of True, use True to replace all nulls.


In [11]:
cols_to_fill = [{'funder': 'unknown'}, {'installer': 'unknown'}, {'subvillage': 'unknown'}, {'public_meeting': True}, {'scheme_management': 'unknown'}, {'scheme_name': 'unknown'}, {'permit': True}]

handle_nulls_inplace(train_values_processed, cols_to_fill)
handle_nulls_inplace(test_values_processed, cols_to_fill)


### Handing Zeros in Numeric Columns
* No change needed
    * latitude : No zeros
    * region_code : No zeros
* Drop data
    * num_private : ~98 of Train,  DROP this COLUMN from Train and Test dataset
    * longitude : ~3% of of Train and ~3% of Test, DROP these ROWS
    * district_code : Very, very few (23 rows in Train and 4 in Test) have zeros. DROP these ROWS.

* Replace Zeros
    * construction_year : ~35% of Train and ~35% of Test - Update zeros with the Average Construction Year.

* Do nothing. These 0 values seem in line with data used on the Offical Tanzanian Water Point Mapping System (WPMS)  [http://wpm.maji.go.tz/]
    * amount_tsh : ~70% of Train were Zero: ??? Take the average of pumps of a similar type? 
    * gps_height : ~ 34% of Train were Zero: ???? Take the average of pumps with similar latitude?
    * population : ~36% of Train and X of Test -  ???


In [12]:
# Functions for handling Zeros in Numeric columns

def count_zeros(df, col_name):
    return df[df[col_name]==0][col_name].count()


In [11]:
numeric_col_names = ['amount_tsh', 'gps_height', 'latitude', 'longitude', 'num_private', 'region_code', 'district_code', 'population', 'construction_year']

for col_name in numeric_col_names:
    the_train_count = count_zeros(train_values_processed, col_name)
    the_test_count = count_zeros(test_values_processed, col_name)
    if(the_test_count + the_train_count > 0):
        print('TRAIN:', col_name, the_train_count)
        print('TEST:', col_name, the_test_count)

TRAIN: amount_tsh 41639
TEST: amount_tsh 10410
TRAIN: gps_height 20438
TEST: gps_height 5211
TRAIN: longitude 1812
TEST: longitude 457
TRAIN: num_private 58643
TEST: num_private 14656
TRAIN: district_code 23
TEST: district_code 4
TRAIN: population 21381
TEST: population 5453
TRAIN: construction_year 20709
TEST: construction_year 5260


In [13]:
# Drop the num_private COLUMN
train_values_processed.drop('num_private', axis=1, inplace=True)
test_values_processed.drop('num_private', axis=1, inplace=True)

In [14]:
# Drop the rows with 0 longitude
train_indices_long = train_values_processed[train_values_processed['longitude'] == 0 ].index
test_indices_long = test_values_processed[test_values_processed['longitude'] == 0 ].index
train_values_processed.drop(train_indices_long, inplace=True)
test_values_processed.drop(test_indices_long, inplace=True)

# drop the rows with 0 district_code
train_indices_distric_code = train_values_processed[train_values_processed['district_code'] == 0 ].index
test_indices_distric_code = test_values_processed[test_values_processed['district_code'] == 0 ].index
train_values_processed.drop(train_indices_distric_code, inplace=True)
test_values_processed.drop(test_indices_distric_code, inplace=True)

In [15]:
# Get the average construction year for TRAIN and TEST
known_const_year_rows = train_values_processed[train_values_processed['construction_year']>0]
avg_counstruction_year = int(known_const_year_rows['construction_year'].mean().round())
print(avg_counstruction_year)

test_known_const_year_rows = test_values_processed[test_values_processed['construction_year']>0]
test_avg_counstruction_year = int(test_known_const_year_rows['construction_year'].mean().round())
print(test_avg_counstruction_year)

1997
1997


In [37]:
# Set construction_year to the average construction year where that value is 0
train_values_processed['construction_year'] = train_values_processed.apply(lambda row: avg_counstruction_year if row['construction_year'] == 0 else row['construction_year'], axis=1)
test_values_processed['construction_year'] = test_values_processed.apply(lambda row: test_avg_counstruction_year if row['construction_year'] == 0 else row['construction_year'], axis=1)

### Normalize String columns - all to lower case

* funder - Who funded the well
* installer - Organization that installed the well
* wpt_name - Name of the waterpoint if there is one
* basin - Geographic water basin
* subvillage - Geographic location
* region - Geographic location
* lga - Geographic location
* ward - Geographic location
* recorded_by - Group entering this row of data
* scheme_management - Who operates the waterpoint
* scheme_name - Who operates the waterpoint
* extraction_type - The kind of extraction the waterpoint uses
* extraction_type_group - The kind of extraction the waterpoint uses
* extraction_type_class - The kind of extraction the waterpoint uses
* management - How the waterpoint is managed
* management_group - How the waterpoint is managed
* payment_type - What the water costs
* payment_type - What the water costs
* water_quality - The quality of the water
* quality_group - The quality of the water
* quantity - The quantity of water
* quantity_group - The quantity of water
* source - The source of the water
* source_type - The source of the water
* source_class - The source of the water
* waterpoint_type - The kind of waterpoint
* waterpoint_type_group - The kind of waterpoin


In [38]:
# Normalize String values function(s)
def normalize_strings(df, col_name):
     df[col_name] = df.apply(lambda row: row[col_name].lower(), axis=1)

In [39]:

string_col_names = ['funder', 'wpt_name', 'basin', 'subvillage', 'region', 'lga', 'ward', 'recorded_by','scheme_management','scheme_name', 'extraction_type', 'extraction_type_group', 'extraction_type_class', 'management', 'management_group', 'payment', 'payment_type', 'water_quality', 'quality_group', 'quantity', 'quantity_group', 'source', 'source_type' , 'source_class', 'waterpoint_type', 'waterpoint_type_group']

for col_name in string_col_names:
    normalize_strings(train_values_processed, col_name)
    normalize_strings(test_values_processed, col_name)


### New Columns/Feature creation
* recorded_year -  Pulling out the year from date_recorded
* waterpoint_age - Calculate as recorded_year - construction_year
* region_with_code - Combine region and region_code. There are more region_code values than region values
* recorded_good_quality - True if quality_group == 'good', False if  anything other than 'good'
* recorded_good_quantity - True if quanity_group == 'sufficient', False if anythign other than 'sufficient'

 

In [40]:
# Functions for creating new features
def get_recorded_year(recorded_date_string):
    year = 0
    date_segs = recorded_date_string.split('-')
    if((len(date_segs) == 3) & (len(date_segs[0]) == 4)):
        try:
            year = int(date_segs[0])
        except:
            print("Not a valid year format.") 
    return year


def get_waterpoint_age(recorded_year, constructed_year):
    age = 0
    is_logical_year = constructed_year > 0
    is_logical_age = recorded_year > constructed_year
    if (is_logical_year & is_logical_age):
        age = recorded_year - constructed_year
    return age


def get_region_name_and_code(region, region_code):
    return region + '_' + str(region_code)


def get_recorded_good_quality(quality_group):
    result = False
    if ('good' == quality_group):
        result = True
    return result


def get_recorded_good_quanity(quanity_group):
    result = False
    if ('enough' == quanity_group):
        result = True
    return result
        

In [41]:
# recorded_year
train_values_processed['recorded_year'] = train_values_processed.apply(lambda row: get_recorded_year(row['date_recorded']), axis=1)
test_values_processed['recorded_year'] = test_values_processed.apply(lambda row: get_recorded_year(row['date_recorded']), axis=1)

In [42]:
# waterpoint_age
train_values_processed['waterpoint_age'] = train_values_processed.apply(lambda row: get_waterpoint_age(row['recorded_year'],row['construction_year']), axis=1)
test_values_processed['waterpoint_age'] = test_values_processed.apply(lambda row: get_waterpoint_age(row['recorded_year'], row['construction_year']), axis=1)

In [43]:
# region_code_and_name
train_values_processed['region_code_and_name'] = train_values_processed.apply(lambda row: get_region_name_and_code(row['region'],row['region_code']), axis=1)
test_values_processed['region_code_and_name'] = test_values_processed.apply(lambda row: get_region_name_and_code(row['region'], row['region_code']), axis=1)

In [44]:
# recorded_good_quality
train_values_processed['recorded_good_quality'] = train_values_processed.apply(lambda row: get_recorded_good_quality(row['quality_group']), axis=1)
test_values_processed['recorded_good_quality'] = test_values_processed.apply(lambda row: get_recorded_good_quality(row['quality_group']), axis=1)

In [45]:
# recorded_good_quanity
train_values_processed['recorded_good_quantity'] = train_values_processed.apply(lambda row: get_recorded_good_quanity(row['quantity_group']), axis=1)
test_values_processed['recorded_good_quantity'] = test_values_processed.apply(lambda row: get_recorded_good_quanity(row['quantity_group']), axis=1)

## Final Prep
* Add the class labels to the TRAINING dataset only
* Save both processed TRAINING and TEST to file for use in EDA and Classifer Modeling


In [46]:
train_values_processed_and_labeled = pd.merge(train_values_processed, train_labels_raw, on='id')

In [47]:
train_values_processed_and_labeled.to_csv('../data/train_processed_labeled.csv', index=False)
test_values_processed.to_csv('../data/test_processed.csv', index=False)

In [48]:
train_values_processed_and_labeled.shape

(57565, 45)

In [50]:
test_values_processed.shape

(14389, 44)

In [51]:
train_values_processed.describe()

id  amount_tsh  gps_height  longitude  latitude  region_code  \
count 57,565.00   57,565.00   57,565.00  57,565.00 57,565.00    57,565.00   
mean  37,108.80      327.78      689.60      35.15     -5.89        15.22   
std   21,454.23    3,044.43      693.57       2.61      2.81        17.86   
min        0.00        0.00      -90.00      29.61    -11.65         1.00   
25%   18,528.00        0.00        0.00      33.28     -8.65         5.00   
50%   37,057.00        0.00      427.00      35.00     -5.17        12.00   
75%   55,669.00       30.00    1,332.00      37.23     -3.37        17.00   
max   74,247.00  350,000.00    2,770.00      40.35     -1.00        99.00   

       district_code  population  construction_year  recorded_year  \
count      57,565.00   57,565.00          57,565.00      57,565.00   
mean            5.73      185.64           1,996.88       2,011.92   
std             9.76      477.83              10.23           0.96   
min             1.00        0.00           1,960.00       2,002.00   
25%             2.00        0.00           1,995.00       2,011.00   
50%             3.00       35.00           1,997.00       2,012.00   
75%             5.00      230.00           2,004.00       2,013.00   
max            80.00   30,500.00           2,013.00       2,013.00   

       waterpoint_age  
count       57,565.00  
mean            15.04  
std             10.26  
min              0.00  
25%              7.00  
50%             14.00  
75%             17.00  
max             53.00

In [53]:
test_values_processed.describe()

id  amount_tsh  gps_height  longitude  latitude  region_code  \
count 14,389.00   14,389.00   14,389.00  14,389.00 14,389.00    14,389.00   
mean  37,164.40      333.17      676.14      35.14     -5.87        15.06   
std   21,346.54    2,550.20      692.07       2.62      2.80        17.46   
min       10.00        0.00      -57.00      29.61    -11.56         1.00   
25%   18,733.00        0.00        0.00      33.24     -8.57         5.00   
50%   37,376.00        0.00      388.00      35.01     -5.18        12.00   
75%   55,768.00       30.00    1,320.00      37.26     -3.37        17.00   
max   74,249.00  200,000.00    2,777.00      40.33     -1.00        99.00   

       district_code  population  construction_year  recorded_year  \
count      14,389.00   14,389.00          14,389.00      14,389.00   
mean            5.73      190.01           1,997.07       2,011.90   
std             9.80      475.79              10.05           0.97   
min             1.00        0.00           1,960.00       2,001.00   
25%             2.00        0.00           1,995.00       2,011.00   
50%             3.00       30.00           1,997.00       2,012.00   
75%             5.00      230.00           2,005.00       2,013.00   
max            80.00   11,469.00           2,013.00       2,013.00   

       waterpoint_age  
count       14,389.00  
mean            14.83  
std             10.09  
min              0.00  
25%              7.00  
50%             14.00  
75%             16.00  
max             53.00